File: 1345-jump-game-iv/1345-jump-game-iv.cpp
class Solution {
public:
    int minJumps(vector<int>& arr) {
        
        int n = arr.size();
        
        unordered_map<int, vector<int>> mp;
        for(int i=0; i<n; i++){
            mp[arr[i]].push_back(i);
        }
        
        vector<bool> visited(n, false);
        
        queue<int> q;
        q.push(0);
        visited[0] = true;
        
        int steps = 0;
        while(!q.empty()){
            
            int size = q.size();
            while(size-- > 0){
                
                int cur = q.front();
                q.pop();
                
                if(cur == n-1)
                    return steps;
                
                int left = cur - 1;
                int right = cur + 1;
                
                if(left >= 0 && !visited[left]){
                    q.push(left);
                    visited[left] = true;
                }
                
                if(right < n && !visited[right]){
                    q.push(right);
                    visited[right] = true;
                }
                
                for(int &idx : mp[arr[cur]]){
                    if(!visited[idx]){
                        q.push(idx);
                        visited[idx] = true;
                    }
                }
                
                mp.erase(arr[cur]);
            }
            
            steps++;
        }
        
        return -1;
    }
};

File: 1345-jump-game-iv/1345-jump-game-iv.py
class Solution:
    def minJumps(self, arr: List[int]) -> int:
        
        maps = {}
        for i in range(len(arr)):
            if arr[i] in maps:  
                new = maps[arr[i]]
                new.append(i)
                maps[arr[i]] = new
            else:
                maps[arr[i]] = [i]
        
        visit = set()
        q = deque()
        q.append(0)
        
        level = 0
        while q:
            size = len(q)
            for i in range(size):
                
                head = q.popleft()
                
                if head == len(arr)-1:
                    return level
                
                if head < 0 or head >= len(arr) or head in visit:
                    continue
                
                if head-1 > 0 and head-1 not in visit:
                    q.append(head - 1)
                
                if head+1 < len(arr) and head+1 not in visit:
                    q.append(head + 1)
                    
                if arr[head] in maps:
                    indexes = maps[arr[head]]
                    for idx in indexes:
                        if idx > 0 and idx < len(arr) and idx not in visit:
                            q.append(idx)
                    
                    del maps[arr[head]]
                
                visit.add(head)
                
            level += 1
        
        return -1
    
        